src/core/iomgr/reachability.h

```diff
@@ -0,0 +1,114 @@
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+#ifndef GRPC_INTERNAL_CORE_IOMGR_REACHABILITY_H
+#define GRPC_INTERNAL_CORE_IOMGR_REACHABILITY_H
+
+// Reachability (as used here) is defined as being the existance of a network
+// route to a specified address. It does not indicate whether a server is
+// serving traffic at that address.
+
+/// Network type: a broad classification of different kinds of network
+/// Useful for systems that want to send traffic only on a restricted class
+typedef enum grpc_network_type {
+  /// Network type is unknown
+  GRPC_NETWORK_UNKNOWN_TYPE,
+  /// Network is a wired connection
+  GRPC_NETWORK_WIRED,
+  /// Network is a wifi connection
+  GRPC_NETWORK_WIFI,
+  /// Network is a cellular connection
+  GRPC_NETWORK_CELLULAR
+} grpc_network_type;
+
+/// A description of a nic
+typedef struct grpc_network_interface {
+  /// Type of the network this nic refers to
+  grpc_network_type network_type;
+  /// Address to pass to bind() to select this nic
+  const sockaddr_t *bind_addr;
+  size_t bind_addr_len;
+  /// A name a sysadmin could use to identify this nic
+  const char *name;
+} grpc_network_interface;
+
+/// Copy a network interface description
+grpc_network_interface *grpc_network_interface_copy(
+    grpc_network_interface *src);
+/// Destroy a network interface description
+void grpc_network_interface_destroy(grpc_network_interface *nic);
+/// Enumerate all network interfaces on this system
+void grpc_enumerate_network_interfaces(grpc_network_interface **nics,
+                                       size_t *nic_count);
+
+/// One time global initialization
+void grpc_reachability_init(void);
+/// Global shutdown - cancels all outstanding watches
+void grpc_reachability_shutdown(void);
+
+/// Watch \a addr for reachability.
+/// If \a addr == NULL, watch for any network being available.
+/// When \a addr becomes reachable, call \a on_reachable with success=true
+/// \a optionally_only_on_nic allows reachability to be awaited only via a
+/// specific network interface
+void grpc_reachability_await_reachable(
+    const sockaddr_t *addr, size_t addr_len,
+    const grpc_network_interface *optionally_only_on_nic,
+    grpc_closure *on_reachable);
+/// Watch \a addr for reachability.
+/// If \a addr == NULL, watch for all networks being unavailable.
+/// When \a addr becomes unreachable, call \a on_unreachable with success=true
+/// \a optionally_only_on_nic allows reachability to be awaited only via a
+/// specific network interface
+void grpc_reachability_await_unreachable(
+    const sockaddr_t *addr, size_t addr_len,
+    const grpc_network_interface *optionally_only_on_nic,
+    grpc_closure *on_unreachable);
+
+/// Query (instantaneous) reachability of \a addr
```

Comment:
Yes, makes sense to drop the query after thinking through this.

On Tue, Feb 23, 2016 at 7:53 AM, Craig Tiller <notifications@github.com>
wrote:

> In src/core/iomgr/reachability.h
> <https://github.com/grpc/grpc/pull/5345#discussion_r53799478>:
>
> > +/// specific network interface
> > +void grpc_reachability_await_reachable(
> > +    const sockaddr_t *addr, size_t addr_len,
> > +    const grpc_network_interface *optionally_only_on_nic,
> > +    grpc_closure *on_reachable);
> > +/// Watch \a addr for reachability.
> > +/// If \a addr == NULL, watch for all networks being unavailable.
> > +/// When \a addr becomes unreachable, call \a on_unreachable with success=true
> > +/// \a optionally_only_on_nic allows reachability to be awaited only via a
> > +/// specific network interface
> > +void grpc_reachability_await_unreachable(
> > +    const sockaddr_t *addr, size_t addr_len,
> > +    const grpc_network_interface *optionally_only_on_nic,
> > +    grpc_closure *on_unreachable);
> > +
> > +/// Query (instantaneous) reachability of \a addr
>
> That's good to know. After sleeping on it, I'm not sure how useful this
> routine is anyway.
>
> Let's look at some use cases:
>
> For name resolution we currently do:
>
> resolve("www.google.com").then(func(addresses) {
>   connectAddresses(addresses);
> });
>
> Once we take into account reachability we'll probably want to:
>
> resolve("www.google.com").then(func(addresses) {
>   set<Address> usable_addresses;
>   func becomeUsable(addr) {
>     usable_addresses.add(addr);
>     publishAddresses(usable_addresses);
>     awaitUnreachable(addr).then(becomeUnusable);
>   }
>   func becomeUnusable(addr) {
>     usable_addresses.remove(addr);
>     publishAddresses(usable_addresses);
>     awaitReachable(addr).then(becomeUsable);
>   }
>   for (addr : addresses) {
>     awaitReachable(addr).then(becomeUsable);
>   }
> });
>
> After shuddering at the complexity a little, I think I'm ok here without
> the query operator - I don't see that it would drastically simplify things
> for us.
>
> We (likely) also want to have something in subchannel to look for
> unreachability post-creation, and to treat that like a disconnect. Here,
> something as simple as:
>
> awaitUnreachable(my_addr, only_on_nic=my_nic).then(disconnect);
>
> is probably all that we need.
>
> So... let's drop the query for now? What do you think @TestUser345
> <https://github.com/TestUser345> - any use case I'm missing?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/grpc/grpc/pull/5345/files#r53799478>.
>